File: LeetCode75/LinkedList/206_reverse_linked_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = NULL;

        while (head != NULL){
            // We will be changing head, in-place, so grab the next iteration
            ListNode* next = head->next;
            // Set the new reversed list's next to the previous value
            head->next = prev;
            // update previous
            prev = head;
            // iterate
            head = next;
        }
        // return last value before null
        return prev;
    }
};

File: LeetCode75/LinkedList/206_reverse_linked_list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseList(ListNode head) {
        ListNode returnVal = null;
        ListNode prevNode = null;

        while (head != null){
            if (returnVal != null){
                prevNode = new ListNode(returnVal.val, returnVal?.next);
            }
            returnVal = new ListNode(head.val, prevNode);

            head = head.next;
        }

        return returnVal;
    }
}